src/control/lib/control/http.go

```diff
@@ -20,13 +21,101 @@ import (
 // didn't set a shorter one on the context.
 const httpReqTimeout = 30 * time.Second
 
+// httpMaxRetries is the number of retries attempted if the HTTP request times out.
+const httpMaxRetries = 5
+
+// HTTPReqTimedOut creates an error indicating that an HTTP request timed out.
+func HTTPReqTimedOut(url string) error {
+	return fmt.Errorf("HTTP request %q: timed out", url)
+}
+
 // httpGetFn represents a function that conforms to the parameters/return values
 // of http.Get
 type httpGetFn func(string) (*http.Response, error)
 
+type httpGetter interface {
+	retryer
+	getURL() *url.URL
+	getBody(context.Context) ([]byte, error)
+}
+
+type httpReq struct {
+	url       *url.URL
+	getFn     httpGetFn
+	getBodyFn func(context.Context, *url.URL, httpGetFn, time.Duration) ([]byte, error)
+}
+
+func (r *httpReq) canRetry(err error, cur uint) bool {
+	if r == nil || r.url == nil {
+		return false
+	}
+
+	if cur >= httpMaxRetries {
+		return false
+	}
+
+	if errors.Cause(err).Error() == HTTPReqTimedOut(r.getURL().String()).Error() {
+		return true
+	}
+
+	return false
+}
+
+func (r *httpReq) onRetry(ctx context.Context, _ uint) error {
+	return nil
+}
+
+func (r *httpReq) retryAfter(_ time.Duration) time.Duration {
+	return time.Second
+}
+
+func (r *httpReq) getRetryTimeout() time.Duration {
+	return httpReqTimeout
+}
+
+func (r *httpReq) getURL() *url.URL {
+	return r.url
+}
+
+func (r *httpReq) httpGetFunc() httpGetFn {
+	if r.getFn == nil {
+		r.getFn = http.Get
+	}
+	return r.getFn
+}
+
+func (r *httpReq) getBody(ctx context.Context) ([]byte, error) {
+	if r.getBodyFn == nil {
+		r.getBodyFn = httpGetBody
+	}
+	return r.getBodyFn(ctx, r.getURL(), r.httpGetFunc(), r.getRetryTimeout())
+}
+
+func httpGetBodyRetry(ctx context.Context, req httpGetter) ([]byte, error) {
+	var result []byte
+	var err error
+	for i := uint(0); ; i++ {
+		result, err = req.getBody(ctx)
+		if err == nil {
+			break
+		}
+
+		if !req.canRetry(err, i) {
+			return nil, err
+		}
+
+		time.Sleep(req.retryAfter(0))
+		if err = req.onRetry(ctx, i); err != nil {
```

Comment:
What is the purpose of this call since it looks like onRetry always returns nil?